### ISSUE TITLE
Decouple piecewise_function evaluation state from function definition to better support multi-threading

### ISSUE THREAD

**Issue by RickBrice** [IfcOpenShell/IfcOpenShell#4727]
I'm using IfcOpenShell v0.8.0, build from source, and BlenderBIM from "Live development environment". For the attached (albeit poorly constructed) model, I am expecting to see IfcAlignmentHorizontal and IfcAlignmentVertical under IfcAlignment as well as a 3D alignment curve in the graphical scene.

Any idea why they aren't there?
[Br01_UTS.ifc.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/15438180/Br01_UTS.ifc.txt)
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/37087370/6bb52a6f-1a9a-4c6c-b6d0-4a27a4a8182e)

**Comment by civilx64** [comment]
To piggyback on the conversation, we would expect to see both the 2D horizontal IfcCompositeCurve in the plane of Z=0 as well as the 3D IfcGradientCurve in the graphical window. This is required to accommodate models where the limits of the vertical alignment are shorter than the extent of the horizontal - which is very common in practice.

**Comment by Moult** [comment]
We load IfcElement by default. Other things are determined on a case by case basis. For example, openings are hidden, space boundaries are lazy loaded when the user requests them, etc. This means that IfcLinearElement is not included in the default project load.

I don't know enough whether we should:

 1. Load IfcLinearElement by default all the time, or...
 2. Not load it by default, but let the user load it later / toggle visibility such as by clicking a bridge segment then pressing a button / hotkey to toggle display modes to show alignments in different ways with cool visualisations (e.g. top down, or automatic elevation annotation text, etc).

My gut feel is the latter - that the user should generally see all "physical" elements loaded by default, and these "special" virtual elements can be toggled on/off later (like space boundaries, voids, structural connections...)

**Comment by RickBrice** [comment]
Being a total BBim novice, @Moult is it fair to summarize your post as BBim doesn’t have support for IfcLinearElement yet (I.e., no UI toggles to display linear elements yet)?

Is there a way to show the elements via a script? What would an example look like?

You can read between the lines and assume that if I get pointed in the right direction, I would like to contribute to bridge and infrastructure capabilities of BBim.

**Comment by Moult** [comment]
Yes I think currently there is no support to load it. The easiest way to load it is to modify the `process_element_filter` function probably in `import_ifc.py` to include not just IfcElement but IfcLinearElement too.

You'll notice the file is named `import_ifc.py`, this is because BBIM started off as a "import" script, not as a native IFC loader. We still have a lot of work to do to refactor the legacy import code into something you can easily call to load in elements lazily.

**Comment by RickBrice** [comment]
I added `+ self.file.by_type("IfcLinearElement")` to https://github.com/IfcOpenShell/IfcOpenShell/blob/90ae7091563348c42623220ea914c42d70485c03/src/blenderbim/blenderbim/bim/import_ifc.py#L423. It didn't work, though I didn't expect the solution to be that simple.

Perhaps this is a project for another day or someone with more experience than me.

**Comment by Moult** [comment]
@aothms I tried testing this file today and nothing seems to come through, as in the iterator never processes anything.

```python
import multiprocessing
import ifcopenshell
import ifcopenshell.geom


ifc_file = ifcopenshell.open('/home/dion/tmp/Br01_UTS.ifc')

print(ifc_file.by_type("IfcElement"))

settings = ifcopenshell.geom.settings()
iterator = ifcopenshell.geom.iterator(settings, ifc_file, multiprocessing.cpu_count())
if iterator.initialize(): # This is False
    while True:
        shape = iterator.get()
        print(shape.id)
        element = ifc_file.by_id(shape.id)
        if not iterator.next():
            break
```

**Comment by aothms** [comment]
Fun stuff, it works on windows, not on linux. 

![afbeelding](https://github.com/IfcOpenShell/IfcOpenShell/assets/1096535/0059f951-78d0-4161-9bcc-2da18539cd54)

I'll need to have a look.

**Comment by aothms** [comment]
Wow, this was a nasty one. I guess it's try when they say be careful with `auto` when using Eigen, but truth be told I still don't understand why it behaved so mean to me.

**Comment by Moult** [comment]
Great stuff! I can see it now, but not for curves:

```
import multiprocessing
import ifcopenshell
import ifcopenshell.geom


# ifc_file = ifcopenshell.open('/home/dion/racbasicsampleproject.ifc')
ifc_file = ifcopenshell.open('/home/dion/tmp/Br01_UTS.ifc')

# print(ifc_file.by_type("IfcElement"))

print(ifcopenshell.ifcopenshell_wrapper.version())
settings = ifcopenshell.geom.settings()
elements = [ifc_file.by_id(133)] # #133=IfcAlignment('0AGLW4kQj3xuDfBO3XsBA9',$,'Ramp B',$,$,#15,#132,$)
print(elements)
iterator = ifcopenshell.geom.iterator(settings, ifc_file, multiprocessing.cpu_count(), include=elements)
if iterator.initialize():
    while True:
        shape = iterator.get()
        print(dir(shape))
        print(dir(shape.geometry))
        break
        element = ifc_file.by_id(shape.id)
        if not iterator.next():
            break
else:
    print('Uhoh') # Uhoh!
```

**Comment by aothms** [comment]
Baby steps, we'll get there. Should be working again. The taxomy is a bit messy wrt to which elements can have a matrix and style and whether that is a requirement to be a top level representation item. Continuously confusing myself.

**Comment by Moult** [comment]
Awesome! Things seem to be working. I _think_ from a user perspective, it makes more sense for the IfcAlignment (IfcLinearPositioningElement) to be visible out of the box, similar to how IfcGrid is visible by default when you load a project. However, I _think_ it's a good idea to _not_ show IfcLinearElements (the hor/vert segments) by default but make them togglable and lazy loaded with a custom UI that perhaps renders it in a nice way (e.g. top down with editing handles and so on).

In any case I just want to check what I'm seeing is right, since some things look suspicious.

Firstly, the IfcAlignment looks good to me. The bridge is along a portion of it. Hooray! However, the segments H1 and H2 seem to be very low down, i.e. it's nearly 300m lower in the Z axis. Is that correct? H1 and H2 are perfectly flat and line up in plan view with the IfcAlignment Ramp B above ... I'm just unsure if it's correct to be so far low down.

![2024-06-27-211958_1078x926_scrot](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/fb7a4adc-3c60-4d27-9b8c-81d8e47c22e6)

The second thing I'd like to confirm is STEP ID 66, IfcAlignmentHorizontal which has no placement or representation. Is that... OK? Intentional? Then also you can see H3 and V4's representations are failing to be processed. I think that's a bug?

![2024-06-27-212119_942x801_scrot](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/a6b9f99f-c287-4352-a1ed-e581578206a4)

Finally V1, V2, and V3 are sort of just really far away from everything else. That seems odd.

BTW I took these screenshots with a false origin in Blender so ignore the XYZs you see in the screenshot.

![2024-06-27-212241_918x518_scrot](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/c269683a-92c1-4680-a63a-00a3c2d4d067)

**Comment by aothms** [comment]
> Oh also the IfcReferent is at the beginning of the IfcAlignment (basically an empty point at that middle left orange dot). Is that correct?

```
>>> f.by_type('IfcReferent')[0].ObjectPlacement.RelativePlacement.Location.DistanceAlong
IfcLengthMeasure(0.)
```

Yes the referent is placed at dist=0

While the sweep is much further along the curve

```
>>> [p.Location.DistanceAlong for p in f.by_type('IFCSECTIONEDSOLIDHORIZONTAL')[0].CrossSectionPositions]
[IfcLengthMeasure(491.416848000023), IfcLengthMeasure(497.939568000031), IfcLengthMeasure(504.462288000039), IfcLengthMeasure(510.985008000047), IfcLengthMeasure(517.507728000055), IfcLengthMeasure(524.030447999947), IfcLengthMeasure(530.553167999955), IfcLengthMeasure(537.075887999963), IfcLengthMeasure(543.598607999971), IfcLengthMeasure(550.121327999979), IfcLengthMeasure(556.644047999987)]
```

> Also I assume intentional, the segments H1 and H2 are at Z = 0 and perfectly flat

> The second thing I'd like to confirm is STEP ID 66, IfcAlignmentHorizontal which has no placement or representation. Is that... OK? Intentional?

Yes. The horizontal part is 2d. The interesting indeed is that in this model the horizontal **alignment** has no representation, but its segments do.

Typically you would expect the horizontal alignment representation to be a chord over the segment representations

```
>>> [r.Representation.Representations[0].Items[0] for r in f[66].IsNestedBy[0].RelatedObjects]
[#45=IfcCurveSegment(.CONTSAMEGRADIENT.,#40,IfcLengthMeasure(0.),IfcLengthMeasure(335.340019232407),#44), #55=IfcCurveSegment(.CONTSAMEGRADIENT.,#50,IfcLengthMeasure(0.),IfcLengthMeasure(-308.389775494132),#54), #65=IfcCurveSegment(.DISCONTINUOUS.,#60,IfcLengthMeasure(0.),IfcLengthMeasure(0.),#64)]
```

Visually this wouldn't change anything.

> Then also you can see H3 and V4's representations are failing to be processed. I think that's a bug?

Yes and no. They are the last segments respectively and the agreement is that there is always a final segment with length=0 because otherwise in certain cases it would not be possible to encode the final tangent. These currently fail to convert mostly because we just don't have the conversion routines setup to retain a representation of a single point (which would be something to work towards in the future).

> Finally V1, V2, and V3 are sort of just really far away from everything else. That seems odd.

These are pretty much useless in 3D but it's not odd. The vertical alignment is a 2D curve that specifies elevation along the horizontal path. Conceptually you can think of an infinite surface of the extrusion of the 2D horizontal curve and drawing the elevation on that surface, but it is represented just on a 2d plane.

> Awesome! Things seem to be working. I think from a user perspective, it makes more sense for the IfcAlignment (IfcLinearPositioningElement) to be visible out of the box, similar to how IfcGrid is visible by default when you load a project. However, I think it's a good idea to not show IfcLinearElements (the hor/vert segments) by default but make them togglable and lazy loaded with a custom UI that perhaps renders it in a nice way (e.g. top down with editing handles and so on).

So all in all. Fully agree. Really exciting to see this becoming a reality. Thanks again @civilx64 and @RickBrice !

**Comment by civilx64** [comment]
> Yes. The horizontal part is 2d. The interesting indeed is that in this model the horizontal alignment has no representation, but its segments do.

This is consistent with nearly all alignment models I've seen in the wild thus far: Representation for IfcAlignment and the segments, but not for IfcAlignmentHorizontal.  AFAIK there is no set guidance on whether this should be considered a typical best practice or has just evolved organically.  You could make the point that the representation of the individual segments is redundant because they are also referenced by the CompositeCurve / GradientCurve / SegmentedReferenceCurve, but maybe there's usefulness that is not currently apparent to me.

**Comment by civilx64** [comment]
> I think it's a good idea to not show IfcLinearElements (the hor/vert segments) by default but make them togglable and lazy loaded with a custom UI that perhaps renders it in a nice way (e.g. top down with editing handles and so on).

@Moult any opinions on utilizing [CAD Sketcher](https://www.cadsketcher.com/) for the alignment authoring UI?  The ability to constrain segments by position / tangency / radius / length would all be part of expected functionality.  It appears to be GPL licensed so maybe as one of multiple options?

Similar question; do you have a good sense for general percentage of the audience that would prefer something more visual-programmy via sverchok?

**Comment by RickBrice** [comment]
I'm on vacation this week so can't participate in this conversation anymore than saying that bridge model and alignment come from my PGSuper software and more likely than not I don't know how to generate a quality model. If anything needs changes, please let me know and all address it in a week or two.

**Comment by civilx64** [comment]
I think everything in the model is great; just some discussions about what should / should not display in Blender when loading an infrastructure model.

**Comment by Moult** [comment]
Awesome! Closing this as things are loading as expected. I'll disable the 2D curves for now until we come up with a nice UI design for it.

CAD Sketcher is an option but I'll need to see more details before knowing how appropriate it is. We've done some experiments with using it for other parametric profiles with some success.

**Comment by RickBrice** [comment]
I'm a little late getting back into the party - Using the latest on the v0.8.0 branch with the Live Development Environment setup and I'm not getting good results in BlenderBIM. The alignment has a sharp jag in it and the deck and girders are no longer aligned. The deck is also too narrow.

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/37087370/cc6c2496-14ca-440b-9605-00ca07b109b2)
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/37087370/b4efeaaf-5a39-4374-88c3-a50e010b82f4)

Something that's completely weird to me - if I select the Geometry and Materials Tab and press the refresh button for Representation, the alignment becomes correct.

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/37087370/ab23b30e-3381-456c-acf5-a9b24414a121)
![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/37087370/46dada40-b144-4c71-8772-9f473fe15af8)

Not sure what to do about this? Perhaps it's just growing pains of being on the bleeding edge.

**Comment by Moult** [comment]
I can't recreate this unfortunately. Is this using the latest IfcOpenBot build?

**Comment by RickBrice** [comment]
No - I compiled IFCOS from source and used the [Live Development Environment setup](https://docs.blenderbim.org/devs/installation.html#live-development-environment).

**Comment by Moult** [comment]
I guess perhaps start by swapping out the ifcopenshell_wrapper dll/so/py files with the ones from the latest bot build https://github.com/IfcOpenBot/IfcOpenShell/commit/c18e4ea6ced759a9b09ef99fb590b5b4d12fd5a0#comments so that we can narrow down the options.

Any messages in Window Menu > Toggle System Console in Blender?

**Comment by RickBrice** [comment]
I replaced my _ifcopenshell_wrapper.pyd with ifc_openshell_wrapper.cp311-win_amd64.pyd and replaced ifcopenshell_wrapper.py from the latest bot build. No change.

No significant messages in the system console.
~~~
Read prefs: "C:\Users\bricer\AppData\Roaming\Blender Foundation\Blender\4.1\config\userpref.blend"
Starting import process :: 0.00
Loading file :: 0.00
Calculate unit scale :: 0.00
Process context filter :: 0.00
Calculate model offset :: 1.30
Predict dense mesh :: 0.00
Set units :: 0.00
Create project :: 0.01
Process element filter :: 0.00
Create styles :: 0.00
Parsing native elements :: 0.00
Done creating geometry
Create native elements :: 0.00
Done creating geometry
0 / 12 elements processed ...
Create elements :: 8.99
Create annotations :: 0.00
Done creating geometry
0 / 1 elements processed ...
Create positioning elements :: 0.07
0 / 13 elements processed ...
Create spatial elements :: 0.00
Create structural items :: 0.00
Create element types :: 0.00
Place objects in collections :: 0.01
Add project to scene :: 0.00
Setting default context :: 0.02
Setup arrays :: 0.06
Import finished in 10.47 seconds
~~~

**Comment by civilx64** [comment]
I am seeing the same behavior with latest bot build and Blender 4.1.  Interestingly if you select the Deck Slab and refresh the representation the shape will be correct but it needs to be manually translated to 0,0,0 to line up with the beams.  Not sure if that is useful for debugging purposes or not.

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/26513472/54e694b9-4d72-4ad6-9456-b3c774ed4d8f)

**Comment by Moult** [comment]
The good news is that I can now recreate this on Windows. It seems to work fine on Linux. I'll see if I can put together a minimal example.

**Comment by RickBrice** [comment]
Kind of a newbie question - what's the difference between the bot build (https://github.com/IfcOpenBot/IfcOpenShell/commit/c18e4ea6ced759a9b09ef99fb590b5b4d12fd5a0#comments) and the blenderbim unstable release (https://github.com/IfcOpenShell/IfcOpenShell/releases)?

What do I do with the bot build zip package if I wanted to use it?

For now, I'm going to take the nuclear option and get everything fresh from github and rebuild.

**Comment by Moult** [comment]
The bot build is a C++ build at a particular point in time for ifcopenshell-python. It represents the latest state of the C++ built for the public to consume.

The BlenderBIM Add-on release is a Python release of blenderbim (which includes ifcopenshell-python) locked at a particular bot build's binaries (typically the latest - but sometimes not if there are showstopper bugs discovered). It represents the latest state of the Python code (which typically moves faster) for the public to consume.

**Comment by Moult** [comment]
OK, I've managed to isolate it. It occurs on Windows with multithreading enabled ("CPU Multiprocessing" is enabled by default in the advanced mode UI when loading). If you disable it, it works again.

**Comment by RickBrice** [comment]
Where is "CPU Multiprocessing" found in the UI so I can disable it?

Could this be a problem that goes all the way down to the C++ level? In IfcCompositeCurve, and IfcCurveSegment can depend on the next IfcCurveSegment. It could be problematic if two adjacent segments are being processed at the same time.

**Comment by Moult** [comment]
It's potentially (likely?) a C++ bug. When loading a file in the file dialog, enable advanced mode. Then the file will not load immediately, but instead show more settings where you can disable CPU multiprocessing: https://docs.blenderbim.org/users/general_usage/ifc_project.html#advanced-mode

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/f4a2d2fc-f276-45f4-a5db-fd0fbc231027)

**Comment by Moult** [comment]
Here is a minimal script to recreate it. There is some Blender code to visually see the results. Alternatively you can delete the bpy code and at the bottom it compares the verts from the two runs and prints the distance they deviate.

![image](https://github.com/IfcOpenShell/IfcOpenShell/assets/88302/8929d52e-5810-4660-98f7-73d500a60b99)

```python
import bpy
import ifcopenshell
import ifcopenshell.geom
import ifcopenshell.util.shape
import numpy as np
import multiprocessing

f = ifcopenshell.open(r'C:\Users\moud308\Downloads\Br01_UTS.ifc')
print('test', f)
e = f.by_id(133)
s = ifcopenshell.geom.settings()
s.set("dimensionality", ifcopenshell.ifcopenshell_wrapper.CURVES_SURFACES_AND_SOLIDS)
it = ifcopenshell.geom.iterator(s, f, multiprocessing.cpu_count(), include=[e])
if it.initialize():
    while True:
        shape = it.get()
        verts = ifcopenshell.util.shape.get_shape_vertices(shape, shape.geometry)
        mesh = bpy.data.meshes.new("a")
        e = shape.geometry.edges
        v = shape.geometry.verts
        vertices = [[v[i], v[i + 1], v[i + 2]] for i in range(0, len(v), 3)]
        edges = [[e[i], e[i + 1]] for i in range(0, len(e), 2)]
        mesh.from_pydata(vertices, edges, [])
        obj = bpy.data.objects.new("with multiprocessing", mesh)
        bpy.context.scene.collection.objects.link(obj)
        if not it.next():
            break
        
f = ifcopenshell.open(r'C:\Users\moud308\Downloads\Br01_UTS.ifc')
print('test', f)
e = f.by_id(133)
s = ifcopenshell.geom.settings()
s.set("dimensionality", ifcopenshell.ifcopenshell_wrapper.CURVES_SURFACES_AND_SOLIDS)
it = ifcopenshell.geom.iterator(s, f, include=[e])
if it.initialize():
    while True:
        shape = it.get()
        verts2 = ifcopenshell.util.shape.get_shape_vertices(shape, shape.geometry)
        mesh = bpy.data.meshes.new("a")
        e = shape.geometry.edges
        v = shape.geometry.verts
        vertices = [[v[i], v[i + 1], v[i + 2]] for i in range(0, len(v), 3)]
        edges = [[e[i], e[i + 1]] for i in range(0, len(e), 2)]
        mesh.from_pydata(vertices, edges, [])
        obj = bpy.data.objects.new("no multiprocessing", mesh)
        bpy.context.scene.collection.objects.link(obj)
        if not it.next():
            break
        
for i, v in enumerate(verts):
    # Is vertex order the same?
    # has_equivalent = False
    # for v2 in verts2:
    #     if np.allclose(v, v2):
    #         has_equivalent = True
    # if has_equivalent:
    #     continue
    # Assuming it's the same...
    if not np.allclose(v, verts2[i]):
        # index, vert1, vert2, distance between them
        print(i, v, verts2[i], np.linalg.norm(verts2[i]-v))
```

**Comment by RickBrice** [comment]
After a clean build of IfcOpenShell, and turning off CPU multiprocessing, the alignment and bridge geometry displays as expected.

**Comment by RickBrice** [comment]
Everything seems to be OK on the C++ side. Using your script for inspiration, I created the following script to use outside of Blender. With and without multiprocessing yield the same results.

I'm not sure where to go from here.

~~~
import ifcopenshell
import ifcopenshell.geom
import ifcopenshell.util.shape
import numpy as np
import multiprocessing
import matplotlib.pyplot as plt

# Initialize the plot
fig1 = plt.figure().add_subplot()

# Open the IFC file and set up geometry settings
f = ifcopenshell.open(r'F:\IfcAlignmentDriver\TestFiles\Br01_UTS.ifc')
print('File opened:', f)
e = f.by_id(133)
s = ifcopenshell.geom.settings()
s.set("dimensionality", ifcopenshell.ifcopenshell_wrapper.CURVES_SURFACES_AND_SOLIDS)

# Process with multiprocessing
cpu_count = multiprocessing.cpu_count()
print('CPU count:', cpu_count)
it = ifcopenshell.geom.iterator(s, f, cpu_count, include=[e])
if it.initialize():
    while True:
        shape = it.get()
        verts = ifcopenshell.util.shape.get_shape_vertices(shape, shape.geometry)
        e = shape.geometry.edges
        v = np.array(shape.geometry.verts).reshape((-1,3))
        x, y, z = v.T
        fig1.plot(x, y, label="multiprocessing")
        if not it.next():
            break

# Process without multiprocessing
f = ifcopenshell.open(r'F:\IfcAlignmentDriver\TestFiles\Br01_UTS.ifc')
print('File reopened:', f)
e = f.by_id(133)
it = ifcopenshell.geom.iterator(s, f, include=[e])
if it.initialize():
    while True:
        shape = it.get()
        verts2 = ifcopenshell.util.shape.get_shape_vertices(shape, shape.geometry)
        v = np.array(shape.geometry.verts).reshape((-1,3))
        x, y, z = v.T
        fig1.plot(x, y, label="no multiprocessing")
        if not it.next():
            break

# Finalize the plot
fig1.legend(loc="upper right")
fig1.axis("square")
fig1.grid(True)
plt.show()  # Display the plot
~~~

![image](https://github.com/user-attachments/assets/c4b9229a-17db-40b4-8e8c-b5db98f7c78d)

**Comment by Moult** [comment]
I can confirm that after swapping out the IfcOpenShell wrapper pyd and py files with the latest bot build, this error no longer occurs.

Is it just luck that a particular bot build contained this potential "race condition / unordered processing" problem or has it actually been fixed?

**Comment by RickBrice** [comment]
I think it is just luck. I've still got the problem in BlenderBIM with IfcOpenShell wrapper pyd compiled from source.

Load with multiprocessing
![image](https://github.com/user-attachments/assets/c3ee427d-6101-4683-b793-99881ec41e21)

Load without multiprocessing
![image](https://github.com/user-attachments/assets/27f74c56-25b4-414b-b37b-0f8a4a48a18b)

I attempted to block race conditions in the geometry caching code, but it didn't seem to have any effect.

Maybe @aothms has some ideas?

~~~
taxonomy::ptr mapping::map(const IfcBaseInterface* inst) {
    auto iden = inst->as<IfcUtil::IfcBaseClass>()->identity();
    if (use_caching_) {
        std::lock_guard<std::mutex> guard(cache_guard_);
        auto it = cache_.find(iden);
        if (it != cache_.end()) {
            return it->second;
        }
    }
    taxonomy::ptr item = nullptr;

    // @todo we should check whether there is a notice performance impact on the large sequence
    // of if-statements and whether a switch on e.g inst->declaration()->index_in_schema()
    // isn't more efficient (which would disable inheritance though).

    bool matched = false;

#include "bind_convert_impl.i"

    if (item) {
        if (use_caching_) {
            std::lock_guard<std::mutex> guard(cache_guard_);
            cache_.insert({iden, item});
        }
    } else if (!matched) {
        Logger::Message(Logger::LOG_ERROR, "No operation defined for:", inst);
    }
    return item;
}
~~~

**Comment by Moult** [comment]
I'm a little confused, are you saying that you cannot recreate it outside blender but you can recreate it within blender? (also did you try with the little script I posted both in and out of blender?)

**Comment by RickBrice** [comment]
You're not confused - that's what I'm saying.

I didn't not run your script within blender. I'm a total newbie with blender and didn't know how to run it. Tell me what I need to do and I'll run the script.

**Comment by Moult** [comment]
https://docs.blender.org/manual/en/latest/editors/text_editor.html

on a train right now  but hopefully that helps. Basically switch to a "Text Editor" panel in Blender (there is also a "Scripting" tab at the top), then create a new text file (or open existing py script) and paste the py script, adjust the path to your file in the code and press the Run Script menu item (or "Play" icon) in the header of the text editor.

That'll create two objects (one with multiprocessing and one without) in the Blender 3d View which you can select in the outliner and use View > Frame Selected in the 3D view to zoom to it.

**Comment by RickBrice** [comment]
Cool - thanks. Learn something new every day.

Still getting the problem. Is there a way to definitively know which IfcOpenShell.pyd is being used? I assume that since I've got the Live Development setup going and when I build and "Install" from VisualStudio, blender is using my most recent build of IfcOpenShell

![image](https://github.com/user-attachments/assets/5cac98ef-e9fa-48b7-a26e-f9f576e3eb00)

**Comment by Moult** [comment]
Sure, in the bottom right of your screenshot there is a little interactive Python console. You can do `import ifcopenshell` then `print(ifcopenshell.__file__)` to see where it is loading the ifcopenshell module from (and therefore which pyd it is loading).

**Comment by RickBrice** [comment]
That's the problem.
![image](https://github.com/user-attachments/assets/c46d6603-7b58-4051-bd6a-fc525f6b8261)

That folder is linked to F:\IfcOpenShell\src\ifcopenshell-python\ifcopenshell which has _ifcopenshell_wrapper.cp311-win_amd64.pyd and _ifcopenshell_wrapper.pyd from July 19, not today.

The install script is putting them in F:\IfcOpenShell\_deps-vs2022-x64-installed\Python311\Lib\site-packages\ifcopenshell.

The good news is, if I copy over the file I compiled today and run your script, both curves come out exactly the same.

I'm going to do some more testing because I still think there is a race condition when adding and retrieving from the cache of mapped instances.

Any idea how to get blender to pick up the latest build of IfcOpenShell that I compile?

**Comment by Moult** [comment]
I'd say just to keep it simple to copy over the *ifcopenshell_wrapper* files to that module.

**Comment by RickBrice** [comment]
@aothms how does the multithreading work with the geometry iterator? I'm specifically thinking if one thread is processing IfcCompositeCurve and it is working through its IfcCurveSegment list, is there a possibility of a race condition where another thread is processing the same IfcCurveSegment but the thread got it from the overall list of geometry to be processed. I really can't explain why multiple threads wasn't working on Windows so I'm grasping at straws, which seems common for multithreading bugs.

**Comment by aothms** [comment]
Sorry, still travelling and need to catch up on a lot of things.

Multi threading issues are hard to debug, but my first starting point to look into would be https://github.com/IfcOpenShell/IfcOpenShell/blob/6a9e76c2afdec067142d60a827541b90b09b416c/src/ifcgeom/taxonomy.cpp#L526-L530

What I would try is not to make these mutable members, but rather `static threadlocal` local variables within the evaluate function.  Because this file contains an alignment and a sweep over it these will be processed consecutively overwriting the member variables even within the scope of the if statement. C++ threadlocal is really quite magical for this purpose.

**Comment by Andrej730** [comment]
There is supposed to be just one, `_ifcopenshell_wrapper.cp311-win_amd64,pyd` takes the priority if it's matching your python version and your platform. 
Before #4686 we use to have just `_ifcopenshell_wrapper.pyd` which didn't allowed to have different platforms/different python version binaries in the same folder. So, I guess `_ifcopenshell_wrapper.pyd` in your case if just from the old times and probably is never used now.

**Comment by RickBrice** [comment]
> What I would try is not to make these mutable members, but rather `static threadlocal` local variables within the evaluate function. 

Tried and failed. `piecewise_function` are sort of composite objects. IfcGradientCurve maps to a pwf with a single span. The span function is a function with lambda functions for horizontal and vertical. When the top level `evaluate(u)` function is called, it caches the first span function of horizontal and calls its `evaluate(u)`. When `evaluate(u)` is called, it finds that the static member `current_span_fn_` has a value so it calls it. The cached `current_span_fn_` is the horizontal lambda function and recursion occurs until there is a stack overflow. 

It's kind of like this where all three instances of `C` operate on `i` through `foo`. All `piecewise_function` instances in a thread share `current_span_fn_`.

~~~
class C
{
    public:
    void foo()  {  static int i = 0;  std::cout << i++ << std::endl;  }
};

int main() {
    C c1,c2, c3;
    c1.foo();
    c2.foo();
    c3.foo();
    return 0;
}

Output: 
0
1
2
~~~

I completely re-implemented `piecewise_function` using the PIMPL idiom so we can play around with it and avoid really long re-compile times. c31c34b7d2b71c206efa1b3917ccaf34bdbffbac

I also implemented critical sections in `mapping::map`. b22ecba17983469d11cd39789a570c5938b9b4c0

At this point, everything seems to be working well. If there is a race condition bug, I'm sure it will show up again. 

> Because this file contains an alignment and a sweep over it these will be processed consecutively overwriting the member variables even within the scope of the if statement

Do you mean that there are multiple threads concurrently calling `piecewise_function::evaluate()` on the same instance of `piecewise_function`? Explain this a little more please, because that would definitely be a big problem.

**Comment by aothms** [comment]
Hm, after I posted my suggestion I was beginning to suspect it was too naive, but thread-local storage requires static or extern definitions, so cannot be applied to ordinary members.

I think maybe we should consider not making `evaluate()` member functions because the way they are setup requires this much state to evaluate them efficiently. Maybe we should instead move these functions to different classes, so that the taxonomy items remain more alike plain old data classes that are immutable (soon..) and easy to cache, serialize, etc.

If we have some `Evaluator` class it can contain both the state and the logic and it would live on the thread currently evaluating the piecewise_function.

> Do you mean that there are multiple threads concurrently calling piecewise_function::evaluate() on the same instance of piecewise_function? Explain this a little more please, because that would definitely be a big problem.

I thought so, but apparently not, due to:

https://github.com/IfcOpenShell/IfcOpenShell/blob/85e022d76ecd41b0c440865395c6efb9e93d9cb1/src/ifcgeom/Iterator.h#L194-L195

In the long run this is what we should work towards. But in the short term I think my suggestion was only a distraction then.

The mapping is currently done file-wide in a single thread (but this I will also change, because I'm currently working on making the IfcFile data structure much more thread friendly).

https://github.com/IfcOpenShell/IfcOpenShell/blob/85e022d76ecd41b0c440865395c6efb9e93d9cb1/src/ifcgeom/Iterator.h#L207

Conversion from taxonomy::item to triangulated shape is performed using multiple threads based on settings:

https://github.com/IfcOpenShell/IfcOpenShell/blob/85e022d76ecd41b0c440865395c6efb9e93d9cb1/src/ifcgeom/Iterator.h#L522

https://github.com/IfcOpenShell/IfcOpenShell/blob/85e022d76ecd41b0c440865395c6efb9e93d9cb1/src/ifcgeom/Iterator.h#L341C9-L342C1

**Comment by RickBrice** [comment]
Alignments seem to be working OK right now - I haven't had a chance to export the Evaluator idea.

**Comment by Moult** [comment]
Oh, should this be closed?

**Comment by RickBrice** [comment]
I change the issue title to what really needs to be done - there is some very good background here. Keep it open for now.

**Comment by RickBrice** [comment]
Resolved via #5344
